### ISSUE TITLE
TensorFlow, NumPy and OpenCV bindings needed for Object Detection Retrain

### ISSUE THREAD

**Issue by sayanshaw24** [SciSharp/TensorFlow.NET#359]
I have been working on writing a final layer retrain for object detection with the YOLOv3 model. I am looking at the following Python example, writing implementation for it in C# on ML.NET, and using TF.NET for python bindings:
https://github.com/YunYang1994/tensorflow-yolov3

I have written some missing bindings already, however the following are some that I need assistance for in order to load label and truth Tensors and compute the confidence, classification and GIOU loss functions for YOLO:
- np: min, max, copy, full, newaxis
- cv2: warpAffine, imread, cvtColor
- tf: train.ExponentialMovingAverage, trainable_variables, no_op(), GraphKeys

For the OpenCV bindings, I found a .NET wrapper called Emgu CV that we could use: http://www.emgu.com/wiki/index.php/Main_Page

Thanks

**Comment by Nucs** [comment]
Hey @sayanshaw24, This weekend we will release `TF.NET` version which will include min, max, copy. 
`newaxis` is tricky with C#'s strong typing.
If it is used in a reshape method - just replace it with a value `1`.
Alternatively you can use `np.expanddim`

Regarding the rest, @Oceania2018

**Comment by Nucs** [comment]
Also I don't think TF.NET have anything to do with `OpenCV` bindings. You might need to find a solution on your own.

**Comment by dgerding** [comment]
Emgu is the long standing rapper of open CV for net. There is an open source or no cost option and then you can move up into commercial versions as well if I recall

Enjoy now. Make it count. or do both.

________________________________

**Comment by sayanshaw24** [comment]
Ok, sounds good, thanks!

Will the new NumSharp nugget include np.full @Nucs? I could just use a duplicate object, but I am worried about issues raising from the pointers having the same memory allocation, but I can try it.

@dgerding Alright, I will start embedding Emgu CV in ML.NET.

**Comment by Nucs** [comment]
>Will the new NumSharp nugget include np.full @Nucs?

Yes.

> I could just use a duplicate object, but I am worried about issues raising from the pointers having the same memory allocation, but I can try it.

If you think you need to perform nd.copy(), then just copy - it is not that performance-expensive

**Comment by Oceania2018** [comment]
@sayanshaw24 Please describe more detail about 
```python
tf: train.ExponentialMovingAverage, trainable_variables, no_op(), GraphKeys
```
It would be better if you can provide the unit test in python or link.

**Comment by sayanshaw24** [comment]
I need the following TensorFlow optimizers/variables in TF.NET so I can use it with ML.NET for object detection retrain:
tf.train.ExponentialMovingAverage, an exponential decay optimizer: https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage

tf.trainable_variables, a list of trainables:
https://www.tensorflow.org/api_docs/python/tf/trainable_variables

tf.no_op, a placeholder operation:
https://www.tensorflow.org/api_docs/python/tf/no_op

tf.GraphKeys, for access to the graph keys, such as updated operations:
https://www.tensorflow.org/api_docs/python/tf/GraphKeys
